runtime/src/snapshot_utils.rs

```diff
@@ -1690,15 +1695,66 @@ pub fn bank_from_snapshot_dir(
     // will calculate and check the accounts hash, so we will still have safety/correctness there.
     bank.set_initial_accounts_hash_verification_completed();
 
-    let timings = BankFromArchiveTimings {
-        rebuild_bank_from_snapshots_us: measure_rebuild.as_us(),
-        full_snapshot_untar_us: measure_build_storage.as_us(),
-        incremental_snapshot_untar_us: 0,
-        verify_snapshot_bank_us: 0,
+    let timings = BankFromDirTimings {
+        rebuild_bank_from_snapshot_us: measure_rebuild.as_us(),
+        build_storage_us: measure_build_storage.as_us(),
     };
     Ok((bank, timings))
 }
 
+/// follow the prototype of fn bank_from_latest_snapshot_archives, implement the from_dir case
+#[allow(clippy::too_many_arguments)]
+pub fn bank_from_latest_snapshot_dir(
+    bank_snapshots_dir: &Path,
+    genesis_config: &GenesisConfig,
+    runtime_config: &RuntimeConfig,
+    account_paths: &[PathBuf],
+    debug_keys: Option<Arc<HashSet<Pubkey>>>,
+    additional_builtins: Option<&Builtins>,
+    account_secondary_indexes: AccountSecondaryIndexes,
+    limit_load_slot_count_from_snapshot: Option<usize>,
+    shrink_ratio: AccountShrinkThreshold,
+    verify_index: bool,
+    accounts_db_config: Option<AccountsDbConfig>,
+    accounts_update_notifier: Option<AccountsUpdateNotifier>,
+    exit: &Arc<AtomicBool>,
+) -> Result<Bank> {
+    info!("Loading bank from snapshot dir");
+    let bank_snapshot = get_highest_bank_snapshot_post(bank_snapshots_dir)
+        .ok_or_else(|| SnapshotError::NoSnapshotSlotDir(bank_snapshots_dir.to_path_buf()))?;
```

Comment:
Yes, we need to look into this more carefully.  It could happen that in the process of adding the account hardlinks the process crashes, so the snapshot dir is incomplete (the state_complete flag file is never created).

In this case, ABS hasn't finish the add_bank_snapshot() call, so it hasn't created the AccountsPackage and send a request to AHV yet, right?  So AHV should not have a chance to process that snapshot dir and convert it into the POST state yet.  

So, any POST snapshot dir should have been completed.

Now the question is: if the incomplete PRE-stage snapshot dir exists due to the process crash, will it ever be purged?

Another new problem is: the accountvec reference counts should not be correct any more because they are kept for only the latest PRE, not the latest POST.

Reply:
OK< I wil move bank_from_latest_snapshot_dir to a new PR.

Reply:
> Another new problem is: the accountvec reference counts should not be correct any more because they are kept for only the latest PRE, not the latest POST.

This is now addressed in https://github.com/solana-labs/solana/pull/31175